backend/agents/Collectors/document_collector.py

```diff
@@ -0,0 +1,214 @@
+
+    def _extract_text_from_url(self, url: str) -> str:
+        # Stream download into a temporary file, then parse.
+        response = requests.get(url, timeout=60)
+        try:
+            response.raise_for_status()
+        except Exception as exc:  # requests.HTTPError
+            raise HTTPException(status_code=500, detail=f"Failed to download {url}: {exc}")
+
+        filename = self._infer_filename_from_url(url, response.headers.get("content-type"))
+        if not self._is_supported(filename):
+            raise HTTPException(
+                status_code=400,
+                detail=(
+                    f"Unsupported file type from URL: {filename}. Supported: {sorted(self.SUPPORTED_EXTENSIONS)}"
+                ),
+            )
+
+        suffix = os.path.splitext(filename)[1]
+        with tempfile.NamedTemporaryFile(delete=True, suffix=suffix) as tmp:
+            tmp.write(response.content)
+            tmp.flush()
+            return self._extract_text_from_file(tmp.name)
+
+    def _extract_text_from_file(self, file_path: str) -> str:
+        _, ext = os.path.splitext(file_path.lower())
+        try:
+            if ext == ".pdf":
+                return self._extract_text_from_pdf(file_path)
+            if ext == ".docx":
+                return self._extract_text_from_docx(file_path)
+            if ext == ".txt":
+                return self._extract_text_from_txt(file_path)
+        except HTTPException:
+            raise
+        except Exception as exc:
+            raise HTTPException(status_code=500, detail=f"Failed to parse {file_path}: {exc}")
+
+        raise HTTPException(status_code=400, detail=f"Unsupported file type: {ext}")
+
+    @staticmethod
+    def _extract_text_from_pdf(file_path: str) -> str:
+        text = pdf_extract_text(file_path) or ""
+        return text.strip()
+
+    @staticmethod
+    def _extract_text_from_docx(file_path: str) -> str:
+        doc = Document(file_path)
+        parts: List[str] = []
+        for paragraph in doc.paragraphs:
+            if paragraph.text and paragraph.text.strip():
+                parts.append(paragraph.text.strip())
+        # Include table text
+        for table in doc.tables:
+            for row in table.rows:
+                cells_text = [cell.text.strip() for cell in row.cells if cell.text and cell.text.strip()]
+                if cells_text:
+                    parts.append("\t".join(cells_text))
+        return "\n".join(parts).strip()
+
+    @staticmethod
+    def _extract_text_from_txt(file_path: str) -> str:
+        with open(file_path, "r", encoding="utf-8", errors="ignore") as f:
+            return f.read().strip()
+
+    def _chunk_text(self, text: str) -> List[str]:
+        if not text:
+            return []
+
+        chunks: List[str] = []
+        for i in range(0, len(text), self.chunk_size):
+            chunk = text[i : i + self.chunk_size]
+            if chunk.strip():
```

Comment:
there should be a package to do all this no ? i know haystack for example and for sure langchain can handle part of it.
what de you think @CedNaru